File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq

def solution(jobs):
	jobs.sort()
	answer = 0
	cur_time = 0
	heap = []
	len_jobs = len(jobs)

	while jobs or heap:
		while jobs and jobs[0][0] <= cur_time:
			start, duration = jobs.pop(0)
			heapq.heappush(heap, (duration, start))

		if heap:
			duration, start = heapq.heappop(heap)
			cur_time += duration
			answer += cur_time - start
		else:
			cur_time = jobs[0][0]

	return answer // len_jobs

print(solution([[0, 3], [1, 9], [2, 6]]))

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> jobs) {
    sort(jobs.begin(), jobs.end());

    int answer = 0;
    int idx = 0;
    int cur_time = 0;
    int jobs_len = jobs.size();

    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;

    while (idx < jobs_len || !pq.empty()) {
        while (idx < jobs_len && jobs[idx][0] <= cur_time) {
            pq.push({jobs[idx][1], jobs[idx][0]});
            idx++;
        }

        if (!pq.empty()) {
            int duration = pq.top().first;
            int start = pq.top().second;
            pq.pop();

            cur_time += duration;
            answer += cur_time - start;
        }
        else {
            cur_time = jobs[idx][0];
        }
    }
    return answer / jobs_len;
}